### PR TITLE
Unify test framework setup

### PR DESCRIPTION
**Pull request by paulidale**
- [x] tests are added or updated

Modify the `register_tests` function so that it returns a boolean indicating if the registration was successful or not.

Add a `cleanup_tests` function that is called after the tests have run to deal with any resource allocations that could have occurred in `register_tests`.

Modify the tests that previously used `test_main` solely to allocate resources and fail gracefully without registering any tests so that they just use `register_tests`.

Unify the test success/fail output into a single function.

This PR is done as two commits, the first changes the function prototype and can be built standalone.  The second modifies existing tests to use the new infrastructure.

### CODE REVIEW

File: test/testutil/driver.c [hunk 1]
```diff
@@ -150,10 +150,28 @@ void set_test_title(const char *title)
     test_title = title == NULL ? NULL : strdup(title);
 }
 
+void test_verdict(int pass, const char *extra, ...)
```
Comment (levitte) [inline test/testutil/driver.c]: I suggest making it `static`...  I can't see that it's needed anywhere else.

File: test/testutil/tu_local.h [hunk 2]
```diff
@@ -10,8 +10,10 @@
 #include <stdlib.h>              /* size_t */
 #include <openssl/bn.h>
 #include <openssl/bio.h>
+#include "../testutil.h"
 
 int subtest_level(void);
+void test_verdict(int verdict, const char *extra, ...) PRINTF_FORMAT(2, 3);
```
Comment (levitte) [inline test/testutil/tu_local.h]: ... and therefore, remove this declaration

File: test/testutil/driver.c (hunk 1 above)
Comment (paulidale) [inline test/testutil/driver.c]: I'm using it in `test_main.c` to indicate an error in registration.
I could change that to a TEST_error or equivalent and make this static.

Comment (paulidale) [inline test/testutil/driver.c]: Which is preferable?

File: test/testutil/test_main.c [hunk 3]
```diff
@@ -9,14 +9,24 @@
...
+    int r;
+
     if (argc > 1)
         test_printf_stderr("Warning: ignoring extra command-line arguments.\n");
 
-    register_tests();
-    return run_tests(argv[0]);
+    if (!register_tests()) {
+        cleanup_tests();
+        test_verdict(0, "registration failed");
```
Comment (levitte) [inline test/testutil/test_main.c]: I would simply remove this line.  From a TAP point of view, the output is wrong anyway, it should be a skip, not an "ok" / "not ok", and furthermore, the tests that still implement their own `test_main` don't output anything when they fail initialization.  Not outputting anything in this case is fine, 'cause the test recipes in `test/recipes` that wrap them will display their own "ok" / "not ok" depending on the exit code of the test program.

**Comment by levitte** [comment]
You know, I'm frankly not entirely sure I like this PR...  `register_tests` was designed to be super simply, just add tests and nothing more, implement `test_main` if you want to make life more complicated.  Nice clean distinction.  With this PR, the distinction between the two becomes a bit more vague.  That in itself makes me wonder if this change is justified.

File: test/testutil/test_main.c (hunk 3 above)
Comment (paulidale) [inline test/testutil/test_main.c]: Okay.

**Comment by paulidale** [comment]
Is the return value from `register_tests` that much complexity?  The added `return 1;` shouldn't be a big burden should it?  The `cleanup_tests` function will typically not be defined.  There are half a dozen tests where `register_tests` was almost but not quite enough.

This is intended as a step in reducing overall complexity.  It should be possible to replace `test_main` entirely without adding any further complexity for the _just adding tests_ case.  The remaining `test_main` tests use argc/argv which is why they haven't been converted.

I also find `test_main` returning zero on success to be a little counter-intuitive.  I understand why it works as it does.

**Comment by levitte** [comment]
No, a returned value isn't that much extra complexity.  However, suddenly *possibly* having to handle an error state when there was *no* need to do so earlier is.

**Comment by levitte** [comment]
Looking at the original comment in testutil, the delineage between `register_tests` and `test_main` is very simple:

```
 * Simple unit tests should implement register_tests().
...
 * Tests that need to perform custom setup or read command-line arguments should
 * implement test_main():
```

With this PR, there's an added "but oh wait, you *can* do some custom setup in `register_tests` after all"...  so suddenly you have two places where you can do custom setup.  That makes it a vague grey area, and that's added complexity.

**Comment by paulidale** [comment]
I'm missing something: handling the return code is done by the test framework and is straightforward.  Users of the framework (i.e. tests) don't see this and several tests are simplified.  Doesn't this reduce overall complexity?

I agree that, at the moment, there is overlap between `test_main` and `register_tests`.  This isn't good but I'm aiming to remove the former.

Thoughts?

**Comment by levitte** [comment]
> I agree that, at the moment, there is overlap between `test_main` and `register_tests`.  This isn't good but I'm aiming to remove the former.

Yeah ok, I get it.  In that case, I would rename `register_tests` to `setup_tests` (not to be mixed up with the internal `setup_test`), 'cause it really isn't about just registering tests.  I guess that the next step will be to pass `argc` and `argv` to `setup_tests`...  so in the end, it's not so much about making a simple change to `register_tests` as it is about splitting `test_main` into a setup part (`setup_tests`) and a cleanup part (`cleanup_tests`) and merging the old `register_tests` into the setup part.  Am I getting that right?  If I am, I'd really rather see the whole change in one PR.

**Comment by paulidale** [comment]
Yeah, that's exactly the idea.
I'll do it as one commit.

I was thinking of adding `test_arg_count` and `test_get_arg` functions instead of passing argc and argv in because passing them in is adding complexity for the tests that don't care about them.  I'd probably also add `test_have_arg` to check if a specific argument was given because most tests are checking for specific arguments only (e.g. _-benchmark_).

Function names subject to approval -- I'm particular bad at choosing meaningful names.

**Comment by levitte** [comment]
I'm not sure I like where this is going.  It becomes added complexity for those who do check the arguments.  I have a hard time seeing that we gain much in the end with those latest ideas...

**Comment by paulidale** [comment]
I've pushed a first stab at argument handling changes -- the function names are horrible and there will be improvements.  Still, I think it simplifies the code for the tests that do use arguments a little.

I'll leave it here until proceed or not is agreed.

File: test/testutil.h [hunk 4]
```diff
@@ -128,8 +127,8 @@
  * rather link to one of the helper main() methods.
  */
 
-/* setup_test() should be called as the first thing in a test main(). */
-void setup_test(void);
+/* setup_test_framework() should be called as the first thing in a test main(). */
+void setup_test_framework(void);
```
Comment (levitte) [inline test/testutil.h]: Maybe rename `finish_test` in a similar manner?

File: test/testutil.h [hunk 5]
```diff
@@ -145,9 +144,14 @@ void set_test_title(const char *title);
 /*
  * Declarations for user defined functions
  */
-void register_tests(void);
-int test_main(int argc, char *argv[]);
-
+int setup_tests(void);
+void cleanup_tests(void);
+
+const char *test_get_program_name(void);
```
Comment (levitte) [inline test/testutil.h]: These should move up, or the comment above will seem to say that these need to be implemented by the user as well.

**Comment by levitte** [comment]
> even though still a bit sceptical.

Don't worry too much about that, though...  I think I mostly need to sleep and wrap my brain around this tomorrow

File: test/testutil.h (hunk 4 above)
Comment (paulidale) [inline test/testutil.h]: Will do.  I think these two can go private.

Comment (paulidale) [inline test/testutil.h]: Not without reworking `afalgtest.c` or putting a new hook in.
Maybe later.

Comment (levitte) [inline test/testutil.h]: I don't quite see why there would be a need for another hook there...

Comment (levitte) [inline test/testutil.h]: Oh, I see now...

**Comment by levitte** [comment]
@paulidale, see attached file for an idea how to fix the afalgtest.c problem.

(and yes, I obviously got used to the idea ;-) )
[register_test.txt](https://github.com/openssl/openssl/files/1164909/register_test.txt)

**Comment by paulidale** [comment]
Thanks for the `afalgtest.c` patch.  That was what I thought would be required.

I'll clean this up early next week.  I'm not sure about the similarity between the `test_get_argument` and `test_has_argument` names.

File: test/testutil/main.c [hunk 6]
```diff
@@ -8,13 +8,86 @@
...
+    return args[n + 1];
+}
 
-    return finish_test(ret);
+size_t test_get_argument_count(void)
+{
+    return arg_count;
 }
+
+int test_has_argument(const char *option)
```
Comment (levitte) [inline test/testutil/main.c]: To get away from the similarity with `test_get_argument`, you could rename this to `test_has_option`

**Comment by paulidale** [comment]
I've removed the WIP tag and squashed some of the earlier commits.
Review time?

File: test/testutil/main.c [hunk 7]
```diff
@@ -35,7 +35,10 @@ int main(int argc, char *argv[])
 {
     int ret = EXIT_FAILURE;
 
-    global_init();
+    if (!global_init()) {
+        test_printf_stderr("Global init failed - aborting\n");
```
Comment (levitte) [inline test/testutil/main.c]: Chicken and egg problem. The framework hasn't been set up yet

Comment (paulidale) [inline test/testutil/main.c]: Doh!

Fixed now.

**Comment by paulidale** [comment]
Rebased to master

File: test/testutil.h [hunk 8]
```diff
@@ -17,34 +17,28 @@
...
+ * The cleanup_tests function will be called even if setup_tests()
+ * returns failure.
  *
- *     // Custom teardown ...
+ * In some cases, early initialization before the framework is set up
+ * may be needed.  In such a case, this should be implemented:
  *
- *     return ret;
- * }
+ * void global_init(void);
```
Comment (levitte) [inline test/testutil.h]: Change this to match the actual return type

File: test/testutil.h [hunk 9]
```diff
@@ -124,30 +118,38 @@
...
+ * test_get_option(const char *) to get an option which includes an
+ *      argument.  NULL is returns if the option is not found.
+ * const char *test_get_program_name(void) returns the name of the test program
+ *      being executed.
  */
+const char *test_get_program_name(void);
+char *test_get_argument(size_t n);
+size_t test_get_argument_count(void);
+int test_has_option(const char *option);
+const char *test_get_option(const char *option);
```
Comment (levitte) [inline test/testutil.h]: I would very much prefer `test_get_option_argument`

File: test/testutil/driver.c [hunk 10]
```diff
@@ -150,10 +146,29 @@ void set_test_title(const char *title)
     test_title = title == NULL ? NULL : strdup(title);
 }
 
+static void test_verdict(int pass, const char *extra, ...) PRINTF_FORMAT(2, 3);
+static void test_verdict(int pass, const char *extra, ...)
```
Comment (levitte) [inline test/testutil/driver.c]: Since it's static, there's no need for the declaration before the definition, just merge these two lines into one.

Comment (paulidale) [inline test/testutil/driver.c]: Force of habit :)
Fixes are in.

**Comment by levitte** [comment]
Ok, looks good to me now.  Will you squash all those commits into one and write a matching commit message, or do you trust our editorial competence? ;-)

**Comment by paulidale** [comment]
I was thinking of squashing them down to a couple rather than one.  The return value and rename of register_tests separately -- I'd like to merge them together into one but the big change between might cause issues..  All the commits after the register_test rename can be combined into one.

If this isn't suitable, I'm happy to mash then into a single commit.  I can scribble some description as well.  I naturally trust the editorial prowess present if someone is keen.

So if anyone has a specific way they'd like to merge this, feel free.
If not, I'll square things up and commit tomorrow (in ten - twelve hours time).

Now we can support test infrastructure that allows generic test related arguments that aren't seen by the test program directly -test-xxxx for a variety of xxxx.  I'm not sure of a use case at present, but it could be handy to keep in reserve.

**Comment by levitte** [comment]
To squash for you was only an offer in case you weren't up for it. You're the author and have priority. As a reviewer, I'm looking for cleanliness, and will therefore encourage that. I'm glad you have a plan, it sounds good to me.

**Comment by paulidale** [comment]
I've squashed this down to a single commit with a new commit message.  Trying to keep the register_test changes separate ended up messier.

I didn't realise it was an offer to squash for me :(

**Comment by paulidale** [comment]
I've rebased this to the current master and added a second commit that fixes the two newly added tests.
Please re-review @levitte

**Comment by paulidale** [comment]
Thanks.

**Pull request closed without merging**